extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergTableWriter.java

```diff
@@ -0,0 +1,557 @@
+
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Stream;
+
+import static io.deephaven.iceberg.base.IcebergUtils.allDataFiles;
+import static io.deephaven.iceberg.base.IcebergUtils.convertToDHType;
+import static io.deephaven.iceberg.base.IcebergUtils.convertToIcebergType;
+import static io.deephaven.iceberg.base.IcebergUtils.createPartitionSpec;
+import static io.deephaven.iceberg.base.IcebergUtils.partitionDataFromPaths;
+import static io.deephaven.iceberg.base.IcebergUtils.verifyWriteCompatibility;
+import static org.apache.iceberg.types.TypeUtil.isPromotionAllowed;
+
+public class IcebergTableWriter {
+
+    /**
+     * The table definition provided by user to write to the Iceberg table. All deephaven tables written by this writer
+     * instance are expected to have this definition.
+     */
+    private final TableDefinition tableDefinition;
+
+    /**
+     * The table definition used for writing the Parquet file. This differs from {@link #tableDefinition} as it:
+     * <ul>
+     * <li>Excludes the partitioning columns</li>
+     * <li>Includes type promotions needed to make {@link #tableDefinition} compatible with the existing table</li>
+     * </ul>
+     */
+    private TableDefinition parquetTableDefinition;
+
+    /**
+     * A one-to-one {@link Map map} from Deephaven column names from the {@link #tableDefinition} to Iceberg field IDs
+     * from the {@link #userSchema}.
+     */
+    private final Map<String, Integer> dhColumnsToIcebergFieldIds;
+
+    /**
+     * Reverse mapping from Iceberg field IDs to Deephaven column names.
+     */
+    private final Map<Integer, String> icebergFieldIdToDhColumn;
+
+    /**
+     * The schema to use when in conjunction with the {@link #dhColumnsToIcebergFieldIds} to map Deephaven columns from
+     * {@link #tableDefinition} to Iceberg columns.
+     */
+    private final Schema userSchema;
+
+    private final IcebergTableAdapter tableAdapter;
+    private final org.apache.iceberg.Table table;
+
+    @Nullable
+    private final TableMetadata tableMetadata;
+    @Nullable
+    private Map<String, Integer> nameMappingDefault;
+
+    private Schema schemaToWrite;
+    private PartitionSpec partitionSpecToWrite;
+
+    private OutputFileFactory outputFileFactory;
+
+    IcebergTableWriter(
+            final TableWriterOptions tableWriterOptions,
+            final IcebergTableAdapter tableAdapter) {
+        this.tableDefinition = tableWriterOptions.tableDefinition();
+        this.dhColumnsToIcebergFieldIds = tableWriterOptions.dhColumnsToIcebergFieldIds();
+        this.icebergFieldIdToDhColumn = new HashMap<>();
+        this.table = tableAdapter.icebergTable();
+        this.userSchema = tableWriterOptions.schema().orElseGet(table::schema);
```

Comment:
It's a bit tricky to find the very initial schema because if someone creates a table, and then does an updateSchema operation, that we don't create a new snapshot but just change the table.schema(). So we cannot be sure what the very first schema was before any snapshots were created.

I have tried to do a basic implementation though, so you can check if it makes sense.